File: modules/core/src/main/java/kr/entree/enderwand/command/Argument.java
package kr.entree.enderwand.command;

import java.util.function.Function;

/**
 * Created by JunHyung Lim on 2020-02-04
 */
public class Argument<T> {
    private final String name;
    private final Function<StringReader, T> parser;

    public Argument(String name, Function<StringReader, T> parser) {
        this.name = name;
        this.parser = parser;
    }

    public T parse(StringReader reader) {
        return parser.apply(reader);
    }
}

File: modules/core/src/main/kotlin/kr/entree/enderwand/command/argument/Argument.kt
package kr.entree.enderwand.command.argument

import kr.entree.enderwand.collection.Reader

/**
 * Created by JunHyung Lim on 2020-01-09
 */
typealias Parser = (Reader<String>) -> Any

typealias Completer = () -> List<String>

data class Argument(
    var description: String = "",
    var optional: Boolean = false,
    private var parser: (Reader<String>) -> Any,
    private var tabCompleter: () -> List<String> = { emptyList() }
) {
    fun parser(parser: (Reader<String>) -> Any) {
        this.parser = parser
    }

    fun tabCompleter(completer: () -> List<String>) {
        this.tabCompleter = completer
    }

    fun parse(reader: Reader<String>) = parser(reader)

    fun tabComplete() = tabCompleter()
}